File: POTD/Two Pointers/Python/3043. Find the Length of the Longest Common Prefix.py
# TIME : O(MAX(M,N))
# SPACE : O(M+N)

class Solution:
    def longestCommonPrefix(self, arr1: List[int], arr2: List[int]) -> int:
        s1=set()
        s2=set()

        for val in arr1:
            s=str(val)

            for i in range(len(s)):
                prefix=s[:i+1]
                s1.add(prefix)

        for val in arr2:
            s=str(val)

            for i in range(len(s)):
                prefix=s[:i+1]
                s2.add(prefix)
        
        maxLen=0

        for s in s1:
            if s in s2:
                maxLen=max(maxLen,len(s))
        
        for s in s2:
            if s in s1:
                maxLen=max(maxLen,len(s))
        
        return maxLen

File: POTD/Two Pointers/Cpp/3043. Find the Length of the Longest Common Prefix.cpp
// using dynamic programming approach

// TIME : o(M*N)
// SPACE : O(M*N) + o(MIN(M,N))
class Solution {
private:
vector<vector<int>> dp;
int getLen(int num1,int num2){
    string s1=to_string(num1);
    string s2=to_string(num2);
    int m=s1.length();
    int n=s2.length();
    int i=0;
    while(i<min(m,n) && s1[i]==s2[i]){
        i+=1;
    }
    return i;
}
int solve(int i,int j,int m,int n,vector<int>&arr1,vector<int>&arr2){
    if(i>=m || j>=n){
        return 0;
    }

    if(dp[i][j]!=-1) return dp[i][j];

    // compare the current
    int res=getLen(arr1[i],arr2[j]);

    res=max(res,solve(i+1,j+1,m,n,arr1,arr2));
    res=max(res,solve(i+1,j,m,n,arr1,arr2));
    res=max(res,solve(i,j+1,m,n,arr1,arr2));

    return dp[i][j]=res;
}
public:
    int longestCommonPrefix(vector<int>& arr1, vector<int>& arr2) {
        int m=arr1.size();
        int n=arr2.size();

        dp.resize(m,vector<int>(n,-1));
        int res=solve(0,0,m,n,arr1,arr2);

        return res;
    }
};
